cudax/include/cuda/experimental/__container/async_mdarray.cuh

```diff
@@ -0,0 +1,901 @@
+  {
+    _CCCL_ASSERT(__n < size(), "cuda::experimental::async_mdarray subscript out of range!");
+    return begin()[__n];
+  }
+
+  //! @brief Returns a reference to the first element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI reference front() noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray front() called on empty async_mdarray!");
+    return begin()[0];
+  }
+
+  //! @brief Returns a reference to the first element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference front() const noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray front() called on empty async_mdarray!");
+    return begin()[0];
+  }
+
+  //! @brief Returns a reference to the last element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI reference back() noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray back() called on empty async_mdarray!");
+    return begin()[size() - 1];
+  }
+
+  //! @brief Returns a reference to the last element of the async_mdarray
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const_reference back() const noexcept
+  {
+    _CCCL_ASSERT(size() != 0, "cuda::experimental::async_mdarray back() called on empty async_mdarray!");
+    return begin()[size() - 1];
+  }
+  //! @}
+
+  //! @addtogroup capacity
+  //! @{
+  //! @brief Returns the current number of elements stored in the async_mdarray.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI size_type size() const noexcept
+  {
+    return __size(__mapping_.extents());
+  }
+
+  //! @brief Returns true if the async_mdarray is empty.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI bool empty() const noexcept
+  {
+    return size() == 0;
+  }
+
+  //! @brief Returns the capacity of the current allocation of the async_mdarray..
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI size_type capacity() const noexcept
+  {
+    return size();
+  }
+
+  //! @brief Returns the maximal size of the async_mdarray.
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI constexpr size_type max_size() const noexcept
+  {
+    return static_cast<size_type>((_CUDA_VSTD::numeric_limits<difference_type>::max)());
+  }
+
+  //! @rst
+  //! Returns a \c const reference to the :ref:`any_resource <cudax-memory-resource-any-resource>`
+  //! that holds the memory resource used to allocate the async_mdarray
+  //! @endrst
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI const __resource_t& get_memory_resource() const noexcept
+  {
+    return __buf_.get_memory_resource();
+  }
+
+  //! @brief Returns the stored stream
+  _CCCL_NODISCARD _CCCL_HIDE_FROM_ABI constexpr ::cuda::stream_ref get_stream() const noexcept
+  {
+    return __buf_.get_stream();
+  }
+
+  //! @brief Replaces the stored stream
+  //! @param __new_stream the new stream
+  //! @note Always synchronizes with the old stream
+  _CCCL_HIDE_FROM_ABI constexpr void change_stream(::cuda::stream_ref __new_stream)
```

Comment:
Changing the stream after construction feels like changing the allocator after construction -- it's not something that one can do to `vector`, and it makes reasoning about code that takes `async_mdarray&` difficult.

Should we consider an alternate design, that makes it easier to move from this `async_mdarray` into a new `async_mdarray` with a different stream?

Reply:
I believe that is different than changing the allocator. The stream is just something that tells the container where future work happens.

Reply:
1. (Re)allocation and future work are coupled through copy assignment.
2. `std::vector` has the invariant that it always has the same allocator throughout its lifetime.
3. If (re)allocation and future work are coupled, then should we also consider adding the invariant that a container always has the same allocator and stream throughout its lifetime?
4. Move construction is relatively cheap and can preserve existing allocations and values.
5. Therefore, should we consider a design in which users can "change" the allocator and/or stream by move-constructing a new container with the desired allocator and/or stream?  This would change the allocator and/or stream for future operations; the current allocation wouldn't necessarily change.

Here's an example of (5).

```c++
env e{allocator, stream}; // whatever the syntax may be
async_mdarray<float, std::dims<2, int>> A{env, num_rows, num_cols};
// ... code ...
async_mdarray<float, std::dims<2, int>> B = /* from somewhere */;
// ... more code ...
env e2{new_allocator, new_stream}; // either or both may change
async_mdarray<float, std::dims<2, int>> A2{e2, std::move(A)}; 
// Assignment from B uses A2's new allocator and new stream.
A2 = B;
```

Reply:
1. yes that is true but I do not see any issue here
2. That is not correct, there is a whole machinery in [allocator_traits](https://en.cppreference.com/w/cpp/memory/allocator_traits) that manages how allocators behave during assignment / construction. If you assign a vector that is not always equal you may need to reallocate 
3. I still dont see how future work and reallocation are coupled. We can happily change streams as long as it is ensured that the streams are properly synchronized
4. We currently only allow the user to change the stream (and the execution policy). As long as every work is in proper stream order there is no reason to put additional constraints in.

In general we *really* want the ability to change the stream of a container, because we never know where we are getting it from. It can realistically happen that we get 3 containers allocated and prepared on different streams and we want to launch an algorithm that takes all of them. In that case we do not only want to synchronize but also ensure that the stream that is internally used is the right one. 

Otherwise every operation on the container would need to take a stream as a required argument and we really do not want that